File: Grind75/57-UniquePaths/cpp/UniquePaths.cpp
#include "UniquePaths.h"
#include "prettyPrint.h"
#include <vector>

using namespace std;

int Solution::uniquePaths(int m, int n) {
  if (m <= 0 || n <= 0)
    return 0;
  vector<int> cache(n);
  for (int r = m-1; r >= 0; r--) {
    for (int c = n-1; c >= 0; c--) {
      if (r == m-1 || c == n-1) {
        cache[c] = 1;
      } else {
        cache[c] = cache[c] + cache[c+1];
      }
    }
  }
  return cache[0];
}

int Solution::uniquePaths2(int m, int n) {
  if (m <= 0 || n <= 0)
    return 0;
  vector<vector<int>> cache(m, vector<int>(n));
  for (int r = m-1; r >= 0; r--) {
    for (int c = n-1; c >= 0; c--) {
      if (r == m-1 || c == n-1) {
        cache[r][c] = 1;
      } else {
        cache[r][c] = cache[r+1][c] + cache[r][c+1];
      }
    }
  }
  prettyPrint2D(cache);
  return cache[0][0];
}

File: Grind75/57-UniquePaths/python/UniquePaths.py
#!/bin/python3

from __future__ import annotations
import sys
sys.path.append('../../Utility/python')
from PrettyPrint import *
from Colors import *
from math import comb

class Solution:
  def uniquePaths(self, m: int, n: int) -> int:
    if m <= 0 or n <= 0:
      return 0
    return comb(m-1 + n-1, m-1)

  def uniquePaths2(self, m: int, n: int) -> int:
    if m <= 0 or n <= 0:
      return 0
    cache: List[int] = [0] * n
    for r in range(m-1, -1, -1):
      for c in range(n-1, -1, -1):
        if r == m-1 or c == n-1:
          cache[c] = 1
        else:
          cache[c] += cache[c+1]
    return cache[0]

if __name__ == '__main__':
  soln = Solution()
  testCases = [
    ( 0, 0, 0 ),
    ( 1, 0, 0 ),
    ( 0, 1, 0 ),
    ( 1, 1, 1 ),
    ( 2, 1, 1 ),
    ( 1, 2, 1 ),
    ( 3, 2, 3 ),
    ( 2, 3, 3 ),
    ( 2, 3, 3 ),
    ( 4, 5, 35 ),
    ( 6, 3, 21 ),
  ]

  for m, n, answer in testCases:
    colorPrint(LIGHT_BLUE, "m: ", m, ",n: ", n, " => ", answer, sep='', end=' ')
    result = soln.uniquePaths(m, n)
    print('-->', colorString(result, GREEN if result == answer else RED))

